helm/prefect-server/templates/_helpers.tpl

```diff
@@ -0,0 +1,262 @@
+{{/*
```

Comment:
There's a lot of magic going on in this file (most of these appear to be borrowed from JupyterHub). While fine, I find the patterns here hard to read and understand what's going on - would prefer a more explicit approach when possible.

To be concrete, take `prefect-server.matchLabels`. The generated values for this template helper change dependent on the file they're used in (and the scope in which they're used), which is different than how most template helpers work (usually they're just shorthand to avoid repeating the same string-generating code everywhere). I had to dig deep to determine that the `matchLabels` used in the apollo deployment/service are different than those in the graphql deployment/service.

I think I'd prefer if the templates here were more explicit. It will mean that service-specific things (like adding the component name to the match labels) will need to be done in each service, but it makes the templates easier to read IMO.

Reply:
re JupyterHub, I copied their styling for docs and used parts of their `componentName` / `nameField` functions but I tried to stay away from duplicating the rest of what they have because I agree it's complicated and they're triple licensed and I didn't want to deal with that.

For the specific example, `prefect-server.matchLabels` could be less magical in a few ways (of increasing verbosity)
1. Extend the docstring to explain that the component name is used, this should be done no matter what
2. Add the componentName to the scope at the top of each template (I do not know how to do this, seems nearly as magical as file name inference)
3. Pass the component name explicitly e.g. `{{- include "prefect-server.matchLabels" (merge (dict "componentName" "hasura") .) -}}`
4. Write a helper for each component that repeats the exact same merge logic but with a hard-coded componentName

4 is what the original PR did and I am pretty against it because then the logic has to be repeatedly changed
3 is okay but the `merge` syntax is pretty verbose and has to be repeated for every item in the deployment/object that contains the component name. I could be convinced though.

I'd argue that adding more documentation to the top of _helpers explaining that we infer component names and doing so consistently is quite clear. I think Helm charts are generally hard to read and I want to make this as clear as possible but I'm not sure any user is actually going to look at how labels are merged, they'll just see that they can define global and component specific labels in the config.

Reply:
Another option (and what I'd recommend) is to have a common set of `matchLabels` used for all deployments, and only add the component name at the end explicitly. Simple, easy to read (I too find the `merge` + `dict` pattern complicated), and doesn't require looking into the template file to understand. This is what I do in dask-gateway and have seen in other charts:

https://github.com/dask/dask-gateway/blob/c2583548df19359d24031e1dd9161c616d3bed50/resources/helm/dask-gateway/templates/gateway/deployment.yaml#L10-L17

Reply:
That is cleaner, I like that. However the component name is being used twice and is composed in the first one e.g.
```
name: prefect-server-apollo
component: apollo
```

https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels is not particularly clear here. Should the `name` be just `apollo`? Should the `component` be `api`?

Reply:
Also note that this is a simple case. What about something like 'annotations' where we want to merge global and local config values?

Reply:
Things like `app.kubernetes.io/name` and `app.kubernetes.io/instance` should refer to the whole application (`prefect-server`), while `app.kubernetes.io/component` should be things like `apollo` or `graphql`. The important thing here is that pods in a deployment have a unique set of `matchLabels` to identify them as different from other deployments.

Reply:
> What about something like 'annotations' where we want to merge global and local config values?

I'd probably drop the global and local config thing, and go for only local for most of these. I can't think of many good use cases for wanting to apply a blanket set of things across all objects - most things a user will want to tweak will be local to a specific component (e.g. add an annotation to pods in the apollo deployment).

That said, I guess what I'm arguing for here is that repeating yourself in each deployment seems clearer to me than doing some magical inferences inside a helper template. There's some readability cost to indirection here that I'm trying to avoid.

Reply:
Thank you everyone for moving this forward! I got it to the state of "works for me" and promptly got absorbed in a machine learning project -- which happily uses the current version. I think @jcrist is on the right track. When I wrote the helpers my concern wasn't so much transparency for the user as that I knew prefect server itself was rapidly evolving -- and I didn't understand the details. The "abstract" helpers and global/local split were meant to allow prefect server itself to change without tracking down duplicated settings. You are all closer to prefect-server itself, and don't have to be so "arms length".

One exception is that -- if you happen to want to run multiple copies of prefect server on one cluster (e.g. for different clients, and/or using different ingresses/load balancers) there are some things that you will want to inject globally -- I think of annotations, in particular. (Maybe affinities/tolerations as well.)

Reply:
> Things like `app.kubernetes.io/name` and `app.kubernetes.io/instance` should refer to the whole application (`prefect-server`), while `app.kubernetes.io/component` should be things like `apollo` or `graphql`.

This is in disagreement with the above K8s docs in which `part-of` refers to the whole application and `name` refers to the name of the part.

> I'd probably drop the global and local config thing, and go for only local for most of these. I can't think of many good use cases for wanting to apply a blanket set of things across all objects - most things a user will want to tweak will be local to a specific component (e.g. add an annotation to pods in the apollo deployment).

I'm not a Helm user but I presumed adding annotations or pull secrets for the whole deployment of our server would be common. I can drop it if that's not the case.

> That said, I guess what I'm arguing for here is that repeating yourself in each deployment seems clearer to me than doing some magical inferences inside a helper template. There's some readability cost to indirection here that I'm trying to avoid.

I think it's worth pointing out that the helper template is only ~200 lines and this deployment is pretty simple. I think repeated logic is harder to maintain, I made these changes because working with it in the original PR was a pain.

I'm not sure what to do here. I think the best solution is to see what it actually looks like / what is needed once it's in use and I'm hesitant to do a bunch more work to simplify it when I may have to add it back in later.